File: python/fgc_host.py
#!/auto/igb-libs/linux/centos/6.x/x86_64/pkgs/python/2.6.8/bin/python
# -*- coding: utf-8 -*-
# host for fgc.exe
# listens to port PORTNUM and do the following
# 0, maintains multiple subprocesses of fgc.exe each loaded
#    some bin files
# 1, listens for verify or match command, defaults to match
#    also listens for input output command of match
# 2, listens for a list of ID
# 3, formats the ID if necessary, get targeted input/output
# 4, If formats specified and found convert as is.
# 5, otherwwise checks the first few words and automatically convert if found
import os, sys, socket, subprocess, time, select, async_subprocess
PORTNUM=47606
HOST=''
FGCDIR="/home/dock/workspace/yuliu/codebase/mycodes/ids/"
BINDIR="/home/dock/workspace/yuliu/codebase/mycodes/ids/bin/"
FGCEXE=FGCDIR+"fgc_proc.exe"
if not os.path.exists(FGCEXE):
    raise IOError

TRANSTABLE={
    u"α":"a",
    u"β":"b",
    u"γ":"g",
    u"δ":"d",
    u"ε":"e",
    u"ζ":"t",
    u"η":"e",
    u"θ":"theta",
    u"κ":"k",
    u"λ":"l",
    u"μ":"mu",
    u"ν":"nu",
    u"ξ":"xi",
    u"Θ":"theta",
    u"π":"pi",
    u"τ":"t",
    u"φ":"phi",
    u"Φ":"phi",
    u"χ":"chi",
    u"ψ":"psi",
    u"Ω":"o",
    u"ω":"o",
    u"_":"",
    u"-":""
}

class fgc(object):
    def __init__(self, binfilename=None, log=None):
        if not binfilename:
            raise IOError
        self.format1=-1;
        self.format2=-1;
        self.set_timeout()
        self.proc=async_subprocess.AsyncPopen([FGCEXE], shell=True,
                                   stdin=subprocess.PIPE, stdout=subprocess.PIPE,
                                   stderr=subprocess.PIPE)
        self.towrite="bind\n"
        self.pollread(True)
        self.towrite=BINDIR+binfilename+"\n"
        self.pollread(True)
        self.ping()
        #make sure process is alive and sane
        return
    def __del__(self):
        self.kill();

    def ping(self):
        """
        returns the format code of the bin bucket
        serves as a sanity check tool
        """
        t1=time.time()
        msgs,_=self.proc.communicate("tell\n")
        while msgs==None:
            t2=time.time()
            if t2-t1>self.timeout:
                sys.exit()
            msgs,_=self.proc.communicate(None)
            time.sleep(1)
            #forcibly exit when sanity is not guaranteed
        msgs=msgs.split("\n")[0].split("\t")
        print msgs
        self.format1,self.format2=msgs[0],msgs[1]
        return (int(msgs[0]), int(msgs[1]))
    def set_timeout(self,timeout=30):
        self.timeout=timeout
    def pollread(self,forerr=False):
        """
        simple polling function
        from nest of heliopolis
        """
        t1=time.time()
        msg,_=self.proc.communicate(self.towrite)
        if msg!=None:
            del self.towrite
            self.towrite=""
            return msg
        else:
            msg=""
            received=False
            cleared=False
            while not (received and cleared):
                t2=time.time()
                if t2-t1>self.timeout:
                    return None
                time.sleep(0.1)
                _msg, err=self.proc.communicate(None)
                if _msg:
                    received=True
                    msg+=_msg
                if received and _msg==None:
                    cleared=True
                if err:
                    print err
                    if forerr:
                        break
            del self.towrite
            self.towrite=""
            return msg
    def pipe (self, another):
        msg=self.pollread();
        if msg==None:
            msg=""
        another._get(msg)
        return another

    def __or__(self, another):
        return self.pipe(another)

    def give (self, msg=None):
        msg=self.pollread()
        if not msg:
            return ""
        else:
            return msg
    def get(self,msg=None):
        self.towrite="\n".join(["DO"]+msg)+"\n\n"
        return

    def _get (self, msg=None):
        self.towrite="DO\n"+msg+"\n\n"
        #pad the message so it goes through
        #note that this does not affect the validate
        #routine
        return

    def verify (self, entries=None):
        """
        issue a validate command with entries
        """
        if len (entries)> 5:
            msg="\n".join(entries[:5])
        else:
            msg="\n".join(entries)

        self.towrite="VALIDATE\nDO\n"+msg+"\n\n"
        print "Trying to verify", msg
        out=self.pollread(True)
        print "got", out
        if out:
            return out.split("\n")
        else:
            return ""

    def kill(self):
        self.proc.communicate("TERMINATETERMINATE\n")
        self.proc.communicate("TERMINATETERMINATE\n")
        self.proc.kill()
        return

class fgc2(object):
    def __init__(self, binfilename=None, fin=subprocess.PIPE, fout=subprocess.PIPE, log=None):
        if not binfilename:
            raise IOError
        self.format1=-1;
        self.format2=-1;
        self.set_timeout()
        self.proc=subprocess.Popen([FGCEXE], shell=True,
                                   stdin=fin, stdout=fout,
                                   stderr=log)
        self.proc.stdin.write("bind\n")
        self.proc.stdin.write(BINDIR+binfilename+"\n")
        #self.ping()
        #make sure process is alive and sane
        return
    def __del__(self):
        self.kill();

    def ping(self):
        """
        returns the format code of the bin bucket
        serves as a sanity check tool
        """
        #read cleanly
        msgs=None
        t1=time.time()
        while (msgs==None):
            time.sleep(1)
            self.proc.stdin.write("tell\n")
            self.proc.stdin.flush()
            msgs=self.pollread()
            t2=time.time()
            if t2-t1>self.timeout:
                break
        if msgs==None:
            sys.exit()
            #forcibly exit when sanity is not guaranteed
        msgs=msgs.split("\n")[0].split("\t")
        self.format1,self.format2=msgs[0],msgs[1]
        return (int(msgs[0]), int(msgs[1]))
    def set_timeout(self,timeout=10):
        self.timeout=timeout
    def pollread(self):
        """
        simple polling function
        from nest of heliopolis
        """
        msg=None
        time.sleep(0.1)
        #we wait a little while for output to be available
        def read():
            poll=select.poll()
            poll.register(self.proc.stdout.fileno(), select.POLLIN or select.POLLPRI)
            fd = poll.poll(self.timeout)
            if len (fd):
                f=fd[0]
                if f[1]>0:
                    return self.proc.stdout.read(1)
        c=read()
        if c != None:
            msg=""
            while c != None:
                msg+=str(c)
                c=read()
        return msg
    def pipe (self, another):
        msg=self.pollread();
        if msg==None:
            msg=""
        another._get(msg)
        return another

    def __or__(self, another):
        return self.pipe(another)

    def give (self, msg=None):
        msg=self.pollread()
        if not msg:
            return ""
        else:
            return msg
    def get(self,msg=None):
        self.proc.stdin.write("\n".join(["DO"]+msg)+"\n\n")
        self.proc.stdin.flush()
        return

    def _get (self, msg=None):
        self.proc.stdin.write("DO\n"+msg+"\n\n")
        #pad the message so it goes through
        #note that this does not affect the validate
        #routine
        self.proc.stdin.flush()
        return

    def verify (self, entries=None):
        """
        issue a validate command with entries
        """
        msg="\n".join(entries)
        self.proc.stdin.write("VALIDATE\nDO\n"+msg+"\n\n")
        out=self.pollread()
        if out:
            return out.split("\n")
        else:
            return ""

    def kill(self):
        self._get("TERMINATETERMINATE\n")
        self._get("TERMINATETERMINATE\n")
        self.proc.kill()
        return

class host(object):
    """
    simple tcp ip host for the converter
    """
    def __init__(self):
        self.entries=[];
        self.exits=[]
        self.entry_uniprot=None
        self.entry_genesym=None
        self.exit_uniprot=None
        self.exit_genesym=None
        self.exit_swissprot=None
        #these three are held in memory running because they
        #are typical exit nodes, if some other conversion is needed
        #that will be done on an on demand basis
        try:
            self.socket=socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        except socket.error as emsg:
            print repr(emsg)
            return
        try:
            self.socket.bind((HOST,PORTNUM))
        except socket.error as emsg:
            print repr(emsg)
            return
        toload=[]
        for fname in os.listdir(BINDIR):
            if ".bin" in fname:
                toload.append(fname)
        print toload
        for some in toload:
            if "entry" in some:
                print "added on demand table", some
                self.entries.append(some)
            elif "exit" in some:
                self.exits.append(some)
        if ("fgc_geneid_genesym.bin" not in toload or
        "fgc_geneid_uniprot.bin" not in toload or
        "fgc_uniprot_geneid.bin" not in toload or
        "fgc_genesym_geneid.bin" not in toload or
        "fgc_swissprot_swissprot.bin" not in toload):
            raise IOError
        else:
            self.exit_uniprot=fgc("fgc_geneid_uniprot.bin")
            self.exit_genesym=fgc("fgc_geneid_genesym.bin")
            self.exit_swissprot=fgc("fgc_swissprot_swissprot.bin")
            self.entry_genesym=fgc("fgc_genesym_geneid.bin")
            self.entry_uniprot=fgc("fgc_uniprot_geneid.bin")
        return;
    def do(self):
        """
        main routine
        """
        self.socket.listen(1)
        self.MSG="--Sending--\n"
        con,add = self.socket.accept()
        con.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        if not con:
            return
        self.data=""
        while True:
            buf=con.recv(1024)
            if not buf:
                break
            self.data+=buf
        self.data=self.data.strip().split("\n")
        commands = self.process_data()
        print len(commands)
        #do something

        #if input is set, grab input if unless it is geneid
        #if not found or input not set, try out
        proc=None
        flag=True
        print self.format1, self.format2
        if self.format1:
            if self.format1=="geneid":
                proc=None
                flag=False
                #no need to pipe
            else:
                if self.format1=="uniprot":
                    proc=self.entry_uniprot
                elif self.format1=="genesym":
                    proc=self.entry_genesym
                for sth in self.entries:
                    if self.format1 in sth:
                        proc=fgc(sth)
                        break
        if flag and not proc:
            if self.guess:
                proc,_=self.tryinput(commands)
            if not proc:
                proc=self.entry_uniprot
                #if nothing is found, use uniprot anyway
        if self.validate:
            if not proc:
                out=self.exit_swissprot.verify(commands)
                self.MSG+="\n".join(out)
                self.MSG+="--Done--"
                con.send(self.MSG)
                return;
            else:
                out=proc.verify(commands)
                self.MSG+="\n".join(out)
                self.MSG+="--Done--"
                con.send(self.MSG)
                return;

        proc2=None
        if self.format2:
            if self.format2=="genesym":
                proc2=self.exit_genesym
            elif self.format2=="swissprot" or self.format2=="uniprot":
                pass;
            else:
                for sth in self.exits:
                    if self.format2 in sth:
                        proc2=fgc(sth)
                        break
        #now if output format is not set, assume swissprot

        #now do the conversion finally!
        if not proc:
            if proc2:
                proc2.get(commands)
                self.MSG+=proc2.give()
            else:
                self.exit_uniprot.get(commands)
                self.MSG+=(self.exit_uniprot|self.exit_swissprot).give()
        else:
            proc.get(commands)
            if proc2:
                self.MSG+=(proc|proc2).give()
            else:
                self.MSG+=(proc|self.exit_uniprot|self.exit_swissprot).give()
        self.MSG+="--Done--"
        con.sendall(self.MSG)
        print proc, proc2
        con.shutdown(socket.SHUT_RDWR)
        con.close()
        return

    def tryinput(self, commands):
        for i in self.entries:
            proc=fgc(i)
            output=proc.verify(commands)
            print "verified", proc, i
            print output
            if len(output)>1:
                return (proc, output)
            else:
                continue;
        return (None, None)

    def __del__(self):
        """
        make sure connection is closed, this should be handled gracefully
        """
        self.entry_genesym.kill()
        self.entry_uniprot.kill()
        self.exit_genesym.kill()
        self.exit_uniprot.kill()
        self.exit_uniprot.kill()
        #self.socket.shutdown()
        self.socket.shutdown(socket.SHUT_RDWR)
        self.socket.close()
        return
    def format (self, commands):
        doform=False
        if "guess" in commands:
            self.guess=True
        if "validate" in commands:
            self.validate=True
        if "from" in commands:
            self.format1=commands[commands.index("from")+1];
            if self.format1=="genesym":
                doform=True
        if "to" in commands:
            self.format2=commands[commands.index("to")+1]
        if "format" in commands:
            doform=True;
            self.format1="genesym"
        if doform:
            #begin format routine
            for j in xrange(1,len(self.data)):
                raw=self.data[j]
                out=""
                u=None
                try:
                    u=raw.decode("utf-8")
                except:
                    try:
                        u=raw.decode("utf-16")
                    except:
                        u=None
                if u:
                    for char in u:
                        if char in TRANSTABLE.keys():
                            o=TRANSTABLE[char]
                        else:
                            o=char.encode("ascii","ignore")
                        out+=o;
                    out="".join([i for i in out if i.isalnum()]).lower()
                    #these are stripped clean to alphanumericals
                else:
                    out="".join([i for i in raw if i.isalnum()]).lower()
                self.data[j]=out;
            self.data[0]=","
        else:
            pass;
    def process_data(self):
        commands= self.data[0].split()
        self.validate=False
        self.guess=False
        self.format1=None
        self.format2=None
        if len(commands)>1:
            self.format(commands)
        else:
            pass
        if not self.format2:
            self.format2="swissprot"
        return self.data
        #we have assembled the input now
        #now simply choose which process to try


if __name__ == "__main__":
    a=host()
    while (True):
        a.do()

File: js/fgc_host.js
//Goals:
//1, implement a non blocking IO manager, this will be trivial
//2, implement a pipe manager to figure out which entry and exit process to call
//3, implement a telnet server, again trivial

subprocess = require ( 'child_process' )
net = require ( 'net' )
http = require ('http')
os = require ( 'os' )
fs = require ( 'fs' )
bf = require ('buffer')
stream = require( 'stream' )
async = require ('async')
//imports
PORTNUM=47606
HOST=''
EXEPATH='/home/dock/workspace/yuliu/codebase/mycodes/ids/fgc_proc.exe'
FORMATEXECPATH='/home/dock/workspace/yuliu/codebase/mycodes/ids/format.py'
BINPATH='/home/dock/workspace/yuliu/codebase/mycodes/ids/bin/'
ENTRIES=[];
MIDDLES=[];
EXITS=[];
TRANSTABLE={
    "α":"a",
    "β":"b",
    "γ":"c",
    "δ":"d",
    "ε":"e",
    "ζ":"t",
    "η":"e",
    "θ":"theta",
    "κ":"k",
    "λ":"l",
    "μ":"mu",
    "ν":"nu",
    "ξ":"xi",
    "Θ":"theta",
    "π":"pi",
    "τ":"t",
    "φ":"phi",
    "Φ":"phi",
    "χ":"chi",
    "ψ":"psi",
    "Ω":"o",
    "ω":"o",
    "alpha":"a",
    "beta":"b",
    "kappa":"k",
    "lambda":"l",
    "delta":"d",
    "gamma":"c",
    "_":"",
    "-":""
}

CALLBACK = function (data){
    console.log('GLOBAL DUMMY CALLBACK', ''+data);
    return
};
fs.exists(EXEPATH, function (exists){
    if (!exists){
        process.exit(-1);
        throw Error("Coluld not find executable!");
    }
})
fs.readdir(BINPATH,function (err, files){
    for (var idx=0; idx<files.length; idx++){
        if (/.bin/.test(files[idx])){
            new fgc(files[idx]);
        }
    }
})


//globals
function fgc (binname) {
    // main class handling the service
    if (!/bin/.test(binname)){
        console.error('Error loading', binname)
        return null;
    }
    var self=this;//???
    var nametmp = binname.split('.')[0].split('_');
    var bound=false;
    this.from = nametmp[1];
    this.to = nametmp[nametmp.length-1];
    //start the process
    var proc= subprocess.execFile(EXEPATH,[],{maxBuffer: 50000*1024});
    this.pid=proc.pid;
    proc.on('error', function(err){
        throw error('Error spawning', proc.pid);
    })
    var __init_proc = function () {
        proc.stdin.write('bind\n');
        proc.stdin.write(BINPATH+binname+'\n');
        var callback = function (data){
            console.log('STDERR:',binname,''+data+'\n');
            if(/BOUND/.test(''+data)){
                console.log('Sucessfully bound', binname);
                bound = true;
                proc.stderr.removeListener('data', callback);
                proc.stdout.setEncoding('utf8')
            }
        }
        proc.stderr.on('data', callback)
    }; //construct once
    __init_proc();

    if (/entry/.test(binname)){
        this.order = 'entry';
        ENTRIES.push(this);
    } else if ( /exit/.test(binname) ){
        this.order = 'exit';
        EXITS.push(this);
    } else {
        this.order = 'middle';
        MIDDLES.push(this);
    } //add to global stores
    console.log('Now added a process, from ' + this.from + ' to '
                + this.to + ' type ' + this.order);

    this.format = function (key){
        key=key.toLowerCase();
        for (var x in TRANSTABLE){
            if ((new RegExp(x)).test(key)){
                key = key.replace(x,TRANSTABLE[x]);
            }
        }
        return key;
    };
    proc.stdout.on('error',function(){
        throw Error("STDOUT ERROR AT " + binname)
    });

    proc.stdin.on('error',function(){
        throw Error("STDIN ERROR AT " + binname)
    });
    this._in = function (msg,callback){
        //block the next attemp
        //console.log('msg is', 'DO\n'+msg+'\n\n')
        //console.log('Calling child at ', binname)
        if (proc.stdin.write('DO\n'+msg+'\n\n')){
        } else {
            console.log('write failed')
            proc.stdin.on('drain', function(){
                processoc.stdin.write('DO\n'+msg+'\n\n')
            })
        }
        proc.stdout.once('data',callback)
        return null;
    };

    this._v = function (msg, callback){
        //console.log('validating at',binname)
        if (proc.stdin.write('VALIDATE\nDO\n'+msg+'\n\n')){
        } else {
            console.log('write failed at ', self.pid)
            proc.stdin.on('drain', function(){
                proc.stdin.write('VALIDATE\nDO\n'+msg+'\n\n');
            })
        }
        proc.stdout.once('data',callback)
        return null;
    };

    this.map = function (target, callback, msg) {
        //console.log('mapping at',binname)
        //console.log('msg is', msg)
        //console.log('target is', target)
        //console.log('callback is', callback)
        //note: this assumes that the source ID of this object matches the msg
        target = typeof target !== 'undefined' ? target : 'human';
        callback = typeof callback !== 'undefined' ? callback : CALLBACK;
        msg = typeof msg !== 'undefined' ? msg : "N/A";
        if (self.from=="genesym") msg=self.format(msg);
        msg=msg.trim().split('\f').join('\n')
        if (target == self.to){
            return self._in(msg, callback);
            //global call back return
        } else if (target == this.from){
            return self._v(msg, callback);
            //global call back return
        } else {
            for (var i=0; i<EXITS.length; i++) {
                out=EXITS[i];
                //console.log('this is ', self.to, 'next is ', out.from)
                if (out.from != self.to) continue;
                if (out.to == target){
                    return self._in(msg,out.map.bind(out, target, callback));
                    break;
                }
            }   //first check if we map to a target
            if (self.order == 'entry') for (var i=0; i< MIDDLES.length; i++) {
                mid=MIDDLES[i];
                //console.log('this is ', self.to, 'next is ', mid.from)
                if (mid.from==self.to)
                    {
                        //console.log('calling a midpoint')
                        return self._in(msg,mid.map.bind(mid, target, callback));
                        break;
                    }
                    // it is now mid's job to do the above
            }
            //if all fails...
            console.log("Could not map", callback);
            return callback('N/A');
        }
    };
}
console.log('begin service at '+ process.pid)
server = net.createServer(function(c){
    console.log('Connection established at ', c.address())
    //c.setEncoding('utf-8')
    //get strings
    var BUF=[];
    len=-1;
    c.on('data',function(data){
        BUF.push(data);
    })
    //get the data
    aTimer= setInterval(function(){
        if (len==BUF.length)
            //if no more buffer gets read
            {
                clearInterval(aTimer);
                c.emit('received')
            }
        else len=BUF.length;
    }, 100);
    //make sure all is got, intermittance is allowed to be 100ms
    //which should be plenty
    c.setTimeout(function(){
        c.emit('end');
    }, 30000)
    //set a global timeout of 30s
    c.once('received',function(){
        var source='uniprot'
        var target='human'
        //default params
        var buffer = bf.Buffer.concat(BUF);
        var lines=buffer.toString()
        lines=lines.split('\n');
        lines.splice(-1,1);//remove FIN pacakge
        console.log('First line is this', lines[0])
        //chop up by lines
        if (/from/.test(lines[0]) || /to/.test(lines[0])){
            var commands=lines[0].split(' ')
            //now go ahead and read the commands
            var idf=commands.indexOf('from'), idt=commands.indexOf('to')
            source = idf==-1 ? source : commands[idf+1];
            target = idt==-1 ? target : commands[idt+1];
            if (target == 'uniprot'){
                if (/mouse/.test(lines[0])) target ='mouse';
                else target = 'human';
            }
            lines.splice(0,1);
        }
        console.log('source target is ', source ,target)
        var total=lines.length;
        c.emit('ready', source, target , lines);
        c.emit('count', total)
    })
    var callback = function (recall, line, data){
        data=data.split('\n')[0].split('\f')[0];
        //truncate output to one only, can be changed
        try{
            c.write(line+','+data+'\n');
        }
        catch (err){
            return recall(Error("Write Error"));
        }
        //the mapping in the middle is blocking
        return recall(null);
        //signals that the run is finished
    }
    c.once ('ready', function (source, target, lines){
        console.log('Initializing one map')
            //pass a callback closure
        var iter = function (line, recall){
            var flag=true;
            if (flag) for (var i=0; i< EXITS.length; i++) {
                var ext=EXITS[i];
                //if we are already at some exit node
                if (ext.from==source && ext.to==target){
                    ext.map(target, callback.bind(null,recall, line), line )
                    flag=false;
                }
            }
            if (flag) for (var i=0; i< MIDDLES.length; i++) {
                var mid=MIDDLES[i];
                //if we are already at some exit node
                if (mid.from==source && mid.to==target){
                    mid.map(target, callback.bind(null,recall, line), line )
                    flag=false;
                }
            }
            if (flag) for (var i=0; i< ENTRIES.length; i++) {
                var ent=ENTRIES[i];
                if (ent.from == source){
                    ent.map(target, callback.bind(null,recall, line), line )
                    flag= false;
                }
            }
            if (flag) {
                callback(recall,line, 'N/A')
            }
        }
        async.eachSeries(lines,iter,function(err){
            if (err==null){
                console.log('\n--Transaction complete--\n')
                c.emit('end')
            }
            else{
                console.log('Error occurred at the connection')
                c.destroy();
                c.emit('end');
            }
        })
    })

    c.on('error',function(err){
        c.emit('end')
    })
    c.once('end', function (){
        console.log('Client disconnected')
        c.end()
    })
    //end the connection immediately when got signal
})
server.listen(PORTNUM,function(){
    console.log('server running')
});
server.on('error', function(e){
    console.log('Error running server')
    console.log('end service')
})
